runtime/src/bank/builtins/prototypes.rs

```diff
@@ -0,0 +1,92 @@
+use {
+    super::core_bpf_migration::{
+        error::CoreBpfMigrationError, CoreBpfMigration, CoreBpfMigrationConfig,
+    },
+    crate::bank::Bank,
+    solana_program_runtime::invoke_context::BuiltinFunctionWithContext,
+    solana_sdk::pubkey::Pubkey,
+};
+
+/// Transitions of built-in programs at epoch boundaries when features are activated.
+pub struct BuiltinPrototype {
+    pub enable_feature_id: Option<Pubkey>,
+    pub core_bpf_migration: Option<CoreBpfMigrationConfig>,
+    pub program_id: Pubkey,
+    pub name: &'static str,
+    pub entrypoint: BuiltinFunctionWithContext,
+}
+
+impl std::fmt::Debug for BuiltinPrototype {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        let mut builder = f.debug_struct("BuiltinPrototype");
+        builder.field("program_id", &self.program_id);
+        builder.field("name", &self.name);
+        builder.field("enable_feature_id", &self.enable_feature_id);
+        builder.field("core_bpf_migration", &self.core_bpf_migration);
+        builder.finish()
+    }
+}
+
+impl BuiltinPrototype {
+    pub(crate) fn migrate_to_core_bpf(&self, bank: &mut Bank) -> Result<(), CoreBpfMigrationError> {
+        if let Some(config) = &self.core_bpf_migration {
+            config.migrate_builtin_to_core_bpf(
+                bank,
+                &self.program_id,
+                CoreBpfMigration::Builtin,
+            )?;
+        }
+        Ok(())
+    }
+}
```

Comment:
So, yes I agree it would make a lot more sense to call from the `config`, but I made this election here to add an additional safeguard that doesn't require you to provide a `program_id`. It's taken directly from the builtin the config lives on.

If we move the call to `CoreBpfMigrationConfig` and store the type within, we'd either need to duplicate the builtin's program ID or require the bank to pass it in, which has some real estate for mistakes.

Reply:
Added this change. I think it's OK. Worth also mentioning that if anyone was to get the `migration_type` wrong, it would still fail gracefully, and the program ID part is in the runtime and shouldn't be changed.